MINOR: Remove sleep calls and ignore annotation from streams upgrade test

bbejeck opened the pull request:
The `StreamsUpgradeTest::test_upgrade_downgrade_brokers`  used sleep calls in the test which led to flaky test performance and as a result, we placed an `@ignore` annotation on the test.  This PR uses log events instead of the `sleep` calls hence we can now remove the `@ignore` setting.

Ran a branch builder https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2146/ which passed with all tests enabled.

Kicked off another branch builder https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2147/

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
MINOR: Remove sleep calls and ignore annotation

bbejeck commented:
ping for review @guozhangwang. @mjsax, and @vvcephei

guozhangwang commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -99,30 +109,43 @@ def test_upgrade_downgrade_brokers(self, from_version, to_version):
+                   err_msg="Broker did not create all topics in 60 seconds ")
 
         self.driver = StreamsSmokeTestDriverService(self.test_context, self.kafka)
+        self.driver.disable_auto_terminate()
```
Why disable auto terminate here?

bbejeck replied:
We want records to run continuously during the test to verify continued processing after the broker upgrade.

bbejeck replied:
EDIT: actually I think this is not needed, I'll remove this and run the system tests agiain.

guozhangwang commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -99,30 +109,43 @@ def test_upgrade_downgrade_brokers(self, from_version, to_version):
         self.processor1.stop()
 
-        node = self.driver.node
-        node.account.ssh("grep -E 'ALL-RECORDS-DELIVERED|PROCESSED-MORE-THAN-GENERATED' %s" % self.driver.STDOUT_FILE, allow_fail=False)
```
Why remove this line?

bbejeck replied:
It's an oversight on my part, I'll add it back.

bbejeck committed:
MINOR: Updates per comments

guozhangwang commented:
@bbejeck lgtm. Could you trigger system test just on this re-enabled case, say 10 times with the updated PR?

bbejeck commented:
> Could you trigger system test just on this re-enabled case, say 10 times with the updated PR?

Done - https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2148/

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -15,14 +15,16 @@
 from kafkatest.services.zookeeper import ZookeeperService
-from kafkatest.version import LATEST_0_10_0, LATEST_0_10_1, LATEST_0_10_2, LATEST_0_11_0, LATEST_1_0, LATEST_1_1, LATEST_2_0, DEV_BRANCH, DEV_VERSION, KafkaVersion
+from kafkatest.version import LATEST_0_10_0, LATEST_0_10_1, LATEST_0_10_2, LATEST_0_11_0, LATEST_1_0, LATEST_1_1, \
+    LATEST_2_0, DEV_BRANCH, DEV_VERSION, KafkaVersion
```
Seems `LATEST_2_1` is missing here? What about other system tests (maybe it's missing there, too)?

bbejeck replied:
No it's not in the system test, but I'm not sure `2.1` is not uploaded to our AWS s3 bucket yet. I'll try to get that loaded, but in the meantime, I'd like this PR to go forward and if need be, do a follow-on PR to update the versions in the test (and other system tests) and update `version.py` and `base.sh` at that time as well.  This applies to similar comments below.

mjsax replied:
Sure. We can address this in a follow up PR.

bbejeck replied:
I've uploaded the new dependencies to AWS so I've added `LATEST_2_1`

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -15,14 +15,16 @@
+    LATEST_2_0, DEV_BRANCH, DEV_VERSION, KafkaVersion
 
 # broker 0.10.0 is not compatible with newer Kafka Streams versions
 broker_upgrade_versions = [str(LATEST_0_10_1), str(LATEST_0_10_2), str(LATEST_0_11_0), str(LATEST_1_0), str(LATEST_1_1), str(LATEST_2_0), str(DEV_BRANCH)]
```
2.1 missing? Similar below for `metadata_3_or_higher_versions` ?

bbejeck replied:
ack fixed

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -50,14 +52,21 @@ def __init__(self, test_context):
         self.leader = None
```
Unrelated to this PR. Nit: update comment in L41/43 above: `Metadata version was bumped in 0.10.1.0` 

It was also dumped in 2.0 release (that also added version probing)

bbejeck replied:
I don't follow, I should add `metadata version dumped in 2.0 release` then?

mjsax replied:
Something like this. Just for completeness (or maybe remove the comment, because we have corresponding variable anyway?)

bbejeck replied:
ack

guozhangwang replied:
`dropped` -> `bumped` ?

bbejeck replied:
ack 
EDIT:
actually from @mjsax comment above (https://github.com/apache/kafka/pull/6046/files#r242487208) I thought we removed the metadata version and went with version probing instead?

Otherwise, I can make the change from `dropped` to `bumped`

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -99,24 +109,39 @@ def test_upgrade_downgrade_brokers(self, from_version, to_version):
         self.zk.start()
```
unrelated nit: fix indentions above

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -99,24 +109,39 @@ def test_upgrade_downgrade_brokers(self, from_version, to_version):
         self.driver = StreamsSmokeTestDriverService(self.test_context, self.kafka)
+        self.driver.disable_auto_terminate()
+
         self.processor1 = StreamsSmokeTestJobRunnerService(self.test_context, self.kafka)
```
nit: rename to `processor` because this test uses only one processor (the numbering is confusing otherwise)

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -99,24 +109,39 @@ def test_upgrade_downgrade_brokers(self, from_version, to_version):
-        time.sleep(15)
 
-        self.perform_broker_upgrade(to_version)
+        self.verify_process_records()
```
Seems we get the monitor here, but start the processor above -- sounds incorrect?

mjsax reviewed (commented):
Thanks for the PR @bbejeck! Some comments/questions.

bbejeck commented:
@mjsax Thanks for the review!
All tests failed for `test_upgrade_downgrade_brokers`, so I'm making some quick changes to get the tests passing then I circle back and address your comments.

bbejeck committed:
MINOR: change processor name, try to run with verification step

bbejeck commented:
Latest system test for just broker upgrades https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2150/.

Just one failure out of 49 tests, but this has all the verification of the `SmokeTest` included, so it's probably another small timing issue to figure out.

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -119,36 +113,37 @@ def test_upgrade_downgrade_brokers(self, from_version, to_version):
-
+                                          err_msg="Never saw output '%s' on" % self.processed_msg + str(processor.node.account))
+        self.driver.wait()
         self.driver.stop()
```
Doesn't `wait()` mean, block until terminated? If yes, no need to cal `stop()` afterwards (would be a no-op)?

bbejeck replied:
This was pre-existing in the test, also used this way in `streams_smoke_test.py`, not sure if required or not but I'd like to leave as is for the moment.

guozhangwang replied:
Hmm.. I think it should be 

```
self.driver.stop()
self.driver.wait()
```

instead as used in other places.

Not sure why the test itself did not fail though, without calling `stop()` the `wait()` call should fail.

ewencp replied:
`wait` should be used if the `Service` is expected to exit naturally based on previous commands:

```
Wait for the service to finish. This only makes sense for tasks with a fixed amount of work to do. For services that generate output, it is only guaranteed to be available after this call returns.
```

You generally do not need to use it in concert with `stop` as `stop` is generally implemented either to synchronously wait (on a clean shutdown via SIGTERM) or kills aggressively enough that you don't actually have to wait (i.e. SIGKILL).

I suspect many of the uses of `wait`/`stop` with these `Service` classes are just redundant -- that is rare or non-existent elsewhere. Use of `wait` elsewhere is mostly limited to things like a console producer/consumer that are known to need to consume a limited amount of data.

guozhangwang replied:
@ewencp thanks for chiming in!

Just double checked the code of `StreamsTestBaseService#stop_node()` is indeed calling with clean_shutdown to true, i.e. to wait for the pid to die out. So the second `wait` call should be removable.

mjsax reviewed (approved):
One clarification question.

LGTM.

bbejeck commented:
rebased from trunk

mjsax commented:
Test timed out. Retest this please.

bbejeck committed:
MINOR: Updates required for bumping available versions for system tests

bbejeck commented:
@mjsax - 

-  updated to include `LATEST_2_1`
-  updated `version.py`, `base.sh`, and `Dockerfile` for `2.1` release and `2.0.1` release
-  updated `settings.gradle`, `dependencie.gradle` for new streams system test module
-  uploaded required kafka, kafka-streams artifacts to S3

\cc @ijuma @ewencp for changes that affect entire project

bbejeck commented:
Kicked off new system test build https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2193/

bbejeck committed:
MINOR: update comment

bbejeck committed:
MINOR: bump version for 2.0.0 to 2.0.1

guozhangwang commented:
@bbejeck Left a follow-up comment, if 2193 passed I think we can merge it after you addressed https://github.com/apache/kafka/pull/6046#discussion_r245159763

bbejeck committed:
fix directory for new versions of streams-test jar

bbejeck commented:
rebased from trunk

bbejeck commented:
Kicked off new branch builder run https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2227/

This run is just for `StreamsUpgradeTest.test_upgrade_downgrade_brokers` so I don't have to wait too long for results

Broker upgrade tests passed 
http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-01-04--001.1546633226--bbejeck--MINOR_remove_sleeps_from_broker_upgrade_tests--23eca5a/report.html

bbejeck committed:
MINOR: remove redundant wait and stop calls

bbejeck commented:
removed redundant `wait` and `stop` calls.  Kicked off another branch builder https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2228/ 
for all tests in `StreamsUpgradeTest`

EDIT adding the `2.1.0` version to the tests is causing failures in `test_metadata_upgrade`

```
TimeoutError: Could not detect Kafka Streams version 2.1.0 ubuntu@worker5
```

Also this error is occurring EDIT found cause of error for this issue so fixing the previous issue should have things fixed.
```
9:27:20 [INFO:2019-01-04 19:27:20,735]: RunnerClient: kafkatest.tests.streams.streams_upgrade_test.StreamsUpgradeTest.test_metadata_upgrade.to_version=2.2.0-SNAPSHOT.from_version=0.10.0.1: Summary: KafkaVersion instance has no attribute 'startswith'
19:27:20 Traceback (most recent call last):
19:27:20   File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/venv/local/lib/python2.7/site-packages/ducktape-0.7.1-py2.7.egg/ducktape/tests/runner_client.py", line 132, in run
19:27:20     data = self.run_test()
19:27:20   File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/venv/local/lib/python2.7/site-packages/ducktape-0.7.1-py2.7.egg/ducktape/tests/runner_client.py", line 185, in run_test
19:27:20     return self.test_context.function(self.test)
19:27:20   File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/venv/local/lib/python2.7/site-packages/ducktape-0.7.1-py2.7.egg/ducktape/mark/_mark.py", line 324, in wrapper
19:27:20     return functools.partial(f, *args, **kwargs)(*w_args, **w_kwargs)
19:27:20   File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/tests/kafkatest/tests/streams/streams_upgrade_test.py", line 228, in test_metadata_upgrade
19:27:20     self.do_stop_start_bounce(p, from_version[:-2], to_version, counter)
19:27:20   File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/tests/kafkatest/tests/streams/streams_upgrade_test.py", line 394, in do_stop_start_bounce
19:27:20     kafka_version_str = self.get_version_string(new_version)
19:27:20   File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/tests/kafkatest/tests/streams/streams_upgrade_test.py", line 325, in get_version_string
19:27:20     if version.startswith("0") or version.startswith("1") \
19:27:20 AttributeError: KafkaVersion instance has no attribute 'startswith'
```

bbejeck committed:
MINOR: fix error for str(VERSION) remove metadata upgrade for 2.1

bbejeck commented:
Kicked off another build https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2229/

I've removed the `metadata_upgrade` for `2.1` temporarily if this branch builder passes I suggest merging this PR as is to get the broker upgrade tests back in and I'll tackle the `metadata-upgrade` in a follow-on PR
\cc @guozhangwang @vvcephei 
WDYT?

bbejeck committed:
MINOR: Add wait back in broker upgrade test, match pattern instead of fixed for numbers

bbejeck commented:
Last build https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2229/
had four failures 1 in `simple_upgrade_downgrade` which seems to be a timing thing and I've addressed this in this update.  

The other three are in the test `upgrade_downgrade_brokers`, and it turns out we need to add back in the `driver.wait()`method.  The previous run of `upgrade_downgrade_brokers` run with `driver.wait()` method, all tests passed http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-01-04--001.1546633226--bbejeck--MINOR_remove_sleeps_from_broker_upgrade_tests--23eca5a/report.html.

The reason we need to add `wait` was described earlier by [this comment from @ewencp ](https://github.com/apache/kafka/pull/6046#discussion_r245155560).  The `StreamsSmokeTest` calls `SmokeTestDriver.generate` to generate a fixed number of records and then calls `SmokeTestDriver.verify` to verify that all records sent were processed as expected, and then exits (https://github.com/apache/kafka/blob/trunk/streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java#L69-L70).  

This test verifies that `SmokeTestDriver.verify` method prints out `ALL-RECORDS-DELIVERED|PROCESSED-MORE-THAN-GENERATED`, but in all failure cases the `SmokeTestDriver` did not finish producing/processing all records as the last statement in the stdout file were `Echo records processed = XXX` as the service was shut down prematurely.

I've kicked off another build https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2230/ which I believe will pass all tests

bbejeck commented:
All tests passed http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-01-06--001.1546752804--bbejeck--MINOR_remove_sleeps_from_broker_upgrade_tests--56a67a5/report.html

bbejeck committed:
MINOR: try using monitor log for driver

bbejeck committed:
MINOR: add blank line

bbejeck committed:
MINOR: improve count of topics

bbejeck commented:
Thinking about this some more and using any sort of `wait` for an arbitrary amount of time is not the best test practice.  I've refactored this PR to use a `monitor` on the `SmokeTestDriver` stdout to finish and print the success message.

Since the change only affects the `test_upgrade_downgrade_brokers` I'll kick off another branch builder for only the broker upgrade tests.

EDIT:  All tests passed http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-01-06--001.1546824140--bbejeck--MINOR_remove_sleeps_from_broker_upgrade_tests--b371f23/report.html

bbejeck committed:
Revert len call, topics_list is a generator

guozhangwang commented:
> Since the change only affects the `test_upgrade_downgrade_brokers` I'll kick off another branch builder for only the broker upgrade tests.

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2231/ passed, but https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2232 has some issue (and it has the same parameters), are you still looking into it?

bbejeck commented:
>https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2231/ passed, but https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2232 has some issue (and it has the same parameters), are you still looking into it?

I needed to revert one change so I aborted 2232 and pushed an update. I kicked off a new branch builder https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2233/ 

All tests passed http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-01-06--001.1546824140--bbejeck--MINOR_remove_sleeps_from_broker_upgrade_tests--b371f23/report.html

guozhangwang commented:
What's the difference between 2231 and 2233?

bbejeck commented:
>What's the difference between 2231 and 2233?

I removed the `self.driver.wait()` call and instead used the more standard approach of 
``` 
with self.driver.node.account.monitor_log(self.driver.STDOUT_FILE) as driver_monitor:
      self.driver.start()
   ....................
   .......................
  driver_monitor.wait_until('ALL-RECORDS-DELIVERED\|PROCESSED-MORE-THAN-GENERATED',
                                      timeout_sec=180,
                                      err_msg="Never saw output '%s' on" % 'ALL-RECORDS-DELIVERED|PROCESSED-MORE-THAN-GENERATED' + str(self.driver.node.account))
```

bbejeck committed:
MINOR: updated per comments

bbejeck commented:
@guozhangwang updated this
